README.md
# Sensu Go Stale Silence Plugin
[![TravisCI Build Status](https://api.travis-ci.com/rgeniesse/sensu-stale-silence-check.svg?branch=master)](https://travis-ci.com/rgeniesse/sensu-stale-silence-check)
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/rgeniesse/sensu-stale-silence-check)

A Sensu Go check plugin for finding stale silence entries

## Installation

Download the latest version of the sensu-stale-silence-check from [releases][1],
or create an executable script from this source.

From the local path of the sensu-stale-silence-check repository:

```
go build -o /usr/local/bin/sensu-stale-silence-check main.go
```

## Configuration

Example Sensu Go definition:

```json
{
  "type": "CheckConfig",
  "api_version": "core/v2",
  "metadata": {
    "namespace": "default",
    "name": "check_for_stale_silence_entries"
  },
  "spec": {
    "command": "sensu-stale-silence-check -H 127.0.0.1 -u admin -p secret -t 604800",
    "subscriptions": [
      "system"
    ],
    "handlers": [
      "slack"
    ],
    "interval": 3600,
    "publish": true
  }
}
```

## Usage

Help:

```
A Sensu Go check plugin for finding stale silence entries

Usage:
  sensu-stale-silence-check [flags]

Flags:
  -h, --help              help for sensu-stale-silence-check
  -H, --host string       The Sensu API host.
  -p, --password string   A Sensu Go user's password.
  -P, --port string       The port the Sensu API is listening on. (default "8080")
  -t, --threshold int     Threshold in seconds to consider a silenced entry stale (default 604800)
  -T, --timeout int       Time in seconds to consider the API unresponsive (default 10)
  -u, --username string   A Sensu Go user with API access.
```

## Contributing

See https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md

[1]: https://github.com/rgeniesse/sensu-stale-silence-check/releases


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## [0.0.12] - 2019-09-27

### Added

- Update readme
- Update plugin description
- Update changelog

## [0.0.11] - 2019-09-27

### Added

- Initial working release


main.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/spf13/cobra"
)

type Auth struct {
	Access_token  string
	Expires_at    int
	Refresh_token string
}

type Metadata struct {
	Name      string `json:"name"`
	Namespace string `json:"namespace"`
}

type Silenced struct {
	Metadata          Metadata
	Expire            int    `json:"expire"`
	Expire_on_resolve bool   `json:"expire_on_resolve"`
	Creator           string `json:"creator"`
	Check             string `json:"check"`
	Subscription      string `json:"subscription"`
	Begin             int    `json:"begin"`
}

var (
	username, password, host, port string
	threshold, timeout             int
)

func configureRootCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "sensu-stale-silence-check",
		Short: "A Sensu Go check plugin for finding stale silence entries",
		RunE:  run,
	}

	cmd.Flags().StringVarP(&username,
		"username",
		"u",
		os.Getenv("SENSU_API_USER"),
		"A Sensu Go user with API access.")
	cmd.MarkFlagRequired("username")

	cmd.Flags().StringVarP(&password,
		"password",
		"p",
		os.Getenv("SENSU_API_PASSWORD"),
		"A Sensu Go user's password.")
	cmd.MarkFlagRequired("password")

	cmd.Flags().StringVarP(&host,
		"host",
		"H",
		os.Getenv("SENSU_API_HOST"),
		"The Sensu API host.")
	cmd.MarkFlagRequired("host")

	cmd.Flags().StringVarP(&port,
		"port",
		"P",
		"8080",
		"The port the Sensu API is listening on.")

	cmd.Flags().IntVarP(&threshold,
		"threshold",
		"t",
		604800,
		"Threshold in seconds to consider a silenced entry stale")

	cmd.Flags().IntVarP(&timeout,
		"timeout",
		"T",
		10,
		"Time in seconds to consider the API unresponsive")

	return cmd
}

func run(cmd *cobra.Command, args []string) error {
	if len(args) != 0 {
		_ = cmd.Help()
		return fmt.Errorf("invalid argument(s) received")
	}

	return nil
}

func getAuthToken() string {

	myauth := Auth{}

	client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
	req, err := http.NewRequest("GET", "http://"+host+":"+port+"/auth", nil)
	req.SetBasicAuth(username, password)
	resp, err := client.Do(req)
	if err != nil {
		log.Fatal(err)
	}
	err2 := json.NewDecoder(resp.Body).Decode(&myauth)

	if err2 != nil {
		log.Fatal(err2)
	}

	return myauth.Access_token
}

func querySilenced(token string, silenced2 *[]Silenced) {

	bearer := "Bearer " + token
	client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
	req, err := http.NewRequest("GET", "http://"+host+":"+port+"/api/core/v2/namespaces/default/silenced", nil)
	req.Header.Add("Authorization", bearer)
	resp, err := client.Do(req)
	if err != nil {
		log.Fatal(err)
	}

	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Fatal(err)
	}
	err2 := json.Unmarshal([]byte(body), silenced2)

	if err2 != nil {
		log.Fatal(err)
	}
}

func checkIfSilencedOld(silenced3 []Silenced) {

	if len(silenced3) > 0 {
		active_entry := false
		for i := 0; i < len(silenced3); i++ {
			n := time.Unix(int64(silenced3[i].Begin), 0)
			duration := time.Since(n)

			// Catch silence entries that will never resolve.
			if int(duration.Seconds()) > threshold && silenced3[i].Expire == int(-1) && !silenced3[i].Expire_on_resolve {
				fmt.Println("A silenced entry " + silenced3[i].Metadata.Name + " has been active since " + n.String())
				active_entry = true
			}

		}

		if !active_entry {
			fmt.Println("Good news nobody, no stale entries found!")
			os.Exit(0)

		} else {
			os.Exit(1)
		}
	} else {
		fmt.Println("Good news nobody, the silenced endpoint is empty!")
		os.Exit(0)
	}
}

func main() {
	rootCmd := configureRootCommand()
	if err := rootCmd.Execute(); err != nil {
		log.Fatal(err.Error())
	}

	token := getAuthToken()
	silenced := []Silenced{}
	querySilenced(token, &silenced)
	checkIfSilencedOld(silenced)

}
